### PR TITLE
STACK-1092 and STACK-1091 changes to support use-rcv-for-resp

### PR DESCRIPTION
**Pull request by ssrinivasan-a10**
Tested with use_rcv_hop_for_resp config option True, False, 0, 1 and Other.

### CODE REVIEW

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+try:
+    from unittest import mock
+except ImportError:
+    import mock
+
+from octavia.common import data_models as o_data_models
+
+from oslo_config import cfg
+
+from a10_octavia.common.data_models import VThunder
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py]: Done.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import a10_octavia.controller.worker.tasks.virtual_port_tasks as task
+from a10_octavia.tests.common import a10constants
+from a10_octavia.tests.unit.base import BaseTaskTestCase
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py]: Done.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+LB = o_data_models.LoadBalancer(id=a10constants.MOCK_LOAD_BALANCER_ID)
+
+
+class TestHandlerVirtualPortTasks(BaseTaskTestCase):
+
+    def setUp(self):
+        super(TestHandlerVirtualPortTasks, self).setUp()
+        imp.reload(task)
+
+    def _mock_conf(self, use_rcv_hop_for_resp, no_dest_nat):
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py]: Done. Thanks. Learnt something new. Had missed this out when checking for examples of mocking CONF.

File: a10_octavia/common/defaults.py [hunk 4]
```diff
@@ -31,5 +31,6 @@
     'templates': None,
     'conn_limit': None,
     'conn_resume': None,
-    'templates': None
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/common/defaults.py:34]: I am still seeing this used in a10_config.py.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        conf.listener.no_dest_nat = no_dest_nat
+        return conf
+
+    def _mock_listener(self, protocol, conn_limit):
+        listener = o_data_models.Listener(id=a10constants.MOCK_LISTENER_ID)
+        listener.protocol = protocol
+        listener.connection_limit = conn_limit
+        return listener
+
+    def test_create_virtual_port_task(self):
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py]: Renamed this to test_create_http_virtual_port_task. The idea was to incorporate multiple test cases in a single method.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        client_mock = mock.Mock()
+        listener_task = task.ListenersCreate()
+        listener_task.axapi_client = client_mock
+        listener_task.execute(LB, [listener], VTHUNDER)
+        args, kwargs = client_mock.slb.virtual_server.vport.create.call_args
+        self.assertTrue(kwargs.get('use_rcv_hop'))
+        self.assertEqual(kwargs.get('conn_limit'), 1000)
+
+    def test_update_virtual_port_task(self):
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py]: Renamed this to test_update_http_virtual_port_task.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    def test_create_virtual_port_task(self):
+        self._mock_conf(True, False)
+        listener = self._mock_listener('HTTP', 1000)
+
+        client_mock = mock.Mock()
+        listener_task = task.ListenersCreate()
+        listener_task.axapi_client = client_mock
+        listener_task.execute(LB, [listener], VTHUNDER)
+        args, kwargs = client_mock.slb.virtual_server.vport.create.call_args
+        self.assertTrue(kwargs.get('use_rcv_hop'))
```
[the remark this answers is not in the document]
Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py:81]: Added assertIn.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hthompson-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py]: I just learned about the utility the other day as well. It's props to @OmkarTelee-A10 for finding it.

File: a10_octavia/common/defaults.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hthompson-a10) [reply a10_octavia/common/defaults.py:34]: Are we still using a10_config.py for anything? Saw it used for db connection, but I'm pretty sure we just remove a10_config.py and defaults.py. Then refactor the db api module to use oslo CONF directly.

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        listener_task.axapi_client = self.client_mock
+        self._set_http_mock_conf(False, False, 'XFF')
+        listener_task.CONF = self.conf
+
+        with mock.patch('a10_octavia.common.openstack_mappings.virtual_port_protocol',
+                        return_value=listener.protocol):
+            listener_task.execute(LB, [listener], VTHUNDER)
+
+        args, kwargs = self.client_mock.slb.virtual_server.vport.update.call_args
+        self.assertIn('use_rcv_hop', kwargs)
```
Comment (hthompson-a10) [inline a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py:101]: We are testing wayyy to much here. Most of this is just things that should be and not edge cases.

Comment (ssrinivasan-a10) [reply a10_octavia/tests/unit/controller/worker/tasks/test_virtual_port_tasks.py:87]: Done. Only testing for use_rcv_hop_for_resp option.
